server/graphql/v2/mutation/PaymentMethodMutations.js

```diff
@@ -50,30 +64,69 @@ const addCreditCard = {
       currency: collective.currency,
       saved: args.isSavedForLater,
       CollectiveId: collective.id,
+      token: args.creditCardInfo.token,
+      data: pick(args.creditCardInfo, ['brand', 'country', 'expMonth', 'expYear', 'fullName', 'funding', 'zip']),
     };
 
-    if (args.creditCardInfo) {
-      const data = pick(args.creditCardInfo, ['brand', 'country', 'expMonth', 'expYear', 'fullName', 'funding', 'zip']);
-      Object.assign(newPaymentMethodData, { token: args.creditCardInfo.token, data });
-    } else if (args.paymentMethod) {
-      Object.assign(newPaymentMethodData, pick(args.paymentMethod, ['data', 'name', 'token']));
-    } else {
-      throw new ValidationFailed('Either creditCardInfo or paymentMethod must be provided');
-    }
-
     let pm = await models.PaymentMethod.create(newPaymentMethodData);
+
     try {
       pm = await setupCreditCard(pm, { collective, user: req.remoteUser });
     } catch (error) {
       if (!error.stripeResponse) {
         throw error;
       }
 
-      // TODO Add support for 3D secure
-      // Currently ignoring the error, which should be ok because we have the `payment.creditcard.confirmation` email.
+      // unsave payment method if saved, we will resave it in confirmCreditCard
+      await pm.update({ saved: false });
+
+      pm.stripeError = {
+        message: error.message,
+        account: error.stripeAccount,
+        response: error.stripeResponse,
+      };
+
+      return { paymentMethod: pm, stripeError: pm.stripeError, shouldBeSaved: args.isSavedForLater };
+    }
+
+    // Success: delete reference to setupIntent and mark again as saved if needed
+    if (pm.data.setupIntent) {
+      delete pm.data.setupIntent;
     }
 
-    return pm;
+    await pm.update({ confirmedAt: new Date(), data: pm.data });
```

Comment:
The `save` flag doesn't seems to be updated as the comment above is suggesting it should be

Reply:
The comment was copied from the other mutation, sorry. The code here is correct. We set `saved: false` in the case of an error, then the resetting happens in `confirmCreditCard`.